pkg/plugins/placementpolicy/core/core.go

```diff
@@ -73,38 +57,114 @@ func (m *PlacementPolicyManager) GetPlacementPolicyForPod(ctx context.Context, p
 	return ppList[0], nil
 }
 
-func (m *PlacementPolicyManager) GetPodsWithLabels(ctx context.Context, podLabels map[string]string) ([]*corev1.Pod, error) {
-	return m.podLister.List(labels.Set(podLabels).AsSelector())
+func (m *PlacementPolicyManager) filterPlacementPolicyList(ppList []*v1alpha1.PlacementPolicy, pod *corev1.Pod) []*v1alpha1.PlacementPolicy {
+	var filteredPPList []*v1alpha1.PlacementPolicy
+	for _, pp := range ppList {
+		labels := pp.Spec.PodSelector.MatchLabels
+		if utils.HasMatchingLabels(pod.Labels, labels) {
+			filteredPPList = append(filteredPPList, pp)
+		}
+	}
+	return filteredPPList
 }
 
-// AnnotatePod annotates the pod with the placement policy.
-func (m *PlacementPolicyManager) AnnotatePod(ctx context.Context, pod *corev1.Pod, pp *v1alpha1.PlacementPolicy, preferredNodeWithMatchingLabels bool) (*corev1.Pod, error) {
-	annotations := map[string]string{}
-	if pod.Annotations != nil {
-		annotations = pod.Annotations
+func (m *PlacementPolicyManager) GetPolicyInfo(pp *v1alpha1.PlacementPolicy) *PolicyInfo {
+	policyNamespace := pp.Namespace
+	policyName := pp.Name
+	corePolicy := pp.Spec.Policy
+
+	namespacePolicies, namespaceExists := m.policies[policyNamespace]
+
+	if namespaceExists {
+		policy, policyExists := namespacePolicies[policyName]
+
+		if policyExists {
+			return policy
+		}
+
+		created := newPolicyInfo(policyNamespace, policyName, corePolicy.Action, corePolicy.TargetSize)
+		m.policies[policyNamespace][policyName] = created
+		return created
 	}
 
-	preference := "false"
-	if preferredNodeWithMatchingLabels {
-		preference = "true"
+	m.policies[policyNamespace] = make(map[string]*PolicyInfo)
+	created := newPolicyInfo(policyNamespace, policyName, corePolicy.Action, corePolicy.TargetSize)
+	m.policies[policyNamespace][policyName] = created
+	return created
```

Comment:
These 3 lines are repeated in the `if namespaceExists` loop when `policyExists` is false. This can be optimized to be done outside the loop by moving `m.policies[policyNamespace] = make(map[string]*PolicyInfo)` to else condition in `if namespaceExists`.